src/common/media.cpp

```diff
@@ -76,12 +77,134 @@ char getByte(GCodeFilter::State *state);
 static char gcode_buffer[MAX_CMD_SIZE + 1]; // + 1 for NULL char
 static GCodeFilter gcode_filter(&getByte, gcode_buffer, sizeof(gcode_buffer));
 static uint32_t media_loop_read = 0;
-static const constexpr uint32_t MEDIA_LOOP_MAX_READ = 4096;
 
 media_state_t media_get_state(void) {
     return media_state;
 }
 
+#define PREFETCH_SIGNAL_START 1
+#define PREFETCH_SIGNAL_STOP  2
+#define PREFETCH_SIGNAL_FETCH 4
+// These buffers are HUGE. We need to rework the prefetcher logic
+// to be more efficient and add compression.
+#define FILE_BUFF_SIZE 5120
+static char prefetch_buff[2][FILE_BUFF_SIZE];
+static char *file_buff;
+static uint32_t file_buff_level;
+static uint32_t file_buff_pos;
+static osThreadId prefetch_thread_id;
+static GCodeFilter::State prefetch_state;
+osMutexDef(prefetch_mutex);
+osMutexId(prefetch_mutex_id);
+
+static void media_prefetch(const void *) {
+    file_buff_level = file_buff_pos = 0;
+    file_buff = prefetch_buff[1];
+    for (;;) {
+        char *back_buff = prefetch_buff[0];
+        GCodeFilter::State bb_state = GCodeFilter::State::Ok;
+        uint32_t back_buff_level = 0;
+        osEvent event;
+
+        file_buff_level = file_buff_pos = 0;
+
+        event = osSignalWait(PREFETCH_SIGNAL_START | PREFETCH_SIGNAL_STOP | PREFETCH_SIGNAL_FETCH, osWaitForever);
```

Comment:
what happens if this loop really "waits forever"?

Reply:
Literally nothing. It's responsibility of the controlling thread to manage the state of this loop. FreeRTOS tasks are meant to never exit their loop. They are meant to either be terminated or run or be suspended forever.